### PR TITLE
Validate application.properties value type

### LINKED ISSUE #33: Validate application.properties: type value
According the type value (ex : int, boolean, other?), validation should be done:

Ex :

 * with quarkus.datasource.enable-metrics which is a boolean

`quarkus.datasource.enable-metrics = XXXX
`

XXXX should be marked as error

 * with `quarkus.datasource.max-size` which is an int

`quarkus.datasource.max-size = XXXX
`

XXXX should be marked as error

### PR DESCRIPTION
**Pull request by xorye**
Fixes #33

This PR provides type validation for enums integer, boolean and float.

image

image

image

@angelozerr , does the ExtendedConfigDescriptionBuildItem.java file from quarkus.ls and quarkus.jdt.core have to be the same? 

Signed-off-by: David Kwon <dakwon@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
> @angelozerr , does the ExtendedConfigDescriptionBuildItem.java file from quarkus.ls and quarkus.jdt.core have to be the same?

Yes please. The idea (in the future) is to create a common project with thoses classes which will be shared between JDT LS extension and Quarkus LS project and other Java LSP client like IJ could use it.

The PR is very promising and I'm sure @jeffmaury will like it.

But it will require some changes (I will review it in detail). But at first your PR should support the @fbricon  comment https://github.com/redhat-developer/quarkus-ls/issues/33#issuecomment-528374597

In other words:

```
quarkus.http.cors=true <-- OK
quarkus.http.cors=blablabla <-- Error
quarkus.http.cors=${blablabla} <-- OK
```

**Comment by xorye** [comment]
> quarkus.http.cors=${blablabla} <-- OK

Ah yes, I forgot to write the test case for this case. I will write one now.

File: quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesDiagnosticsTest.java [hunk 1]
```diff
@@ -32,17 +31,17 @@
 	@Test
 	public void validateUnknownProperties() throws BadLocationException {
 		String value = "# quarkus.datasource.url=jdbc:postgresql:quarkus_test\n" + //
-				"quarkus.datasource.driver=org.postgresql.Driver\n" + //
-				"quarkus.datasource.username=quarkus_test\n" + //
-				"quarkus.datasource.password=quarkus_test\n" + //
+				"quarkus.datasource.driver=\"org.postgresql.Driver\"\n" + //
```
Comment (xorye) [inline quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesDiagnosticsTest.java]: My mistake, I must remove the quotations for these test cases.

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java [hunk 2]
```diff
@@ -79,7 +79,7 @@ private void validateProperty(Property property) {
 			if (metadata == null) {
 				// Validate Unknown property
 				validateUnknownProperty(propertyName, property);
-			} else {
+			} else if (property.getValue() != null) {
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java]: please remove this if condition and do the test inside validatePropertyValue which must manage validation type (as your PR) and required validation

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java [hunk 3]
```diff
@@ -130,13 +130,61 @@ private void validatePropertyValue(String propertyName, ExtendedConfigDescriptio
 			return;
 		}
 
-		// TODO validate boolean, int, enums types
-		String type = metadata.getType();
+		String value = property.getValue().getText().trim();
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java]: Replace with 

```java
String value = property.getPropertyValue();
if (value == null || value.isEmpty()) {
	// TODO: validate required.
	return;
}
```

and trim (if required should be done in property.getPropertyValue())

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java [hunk 4]
```diff
@@ -130,13 +130,61 @@ private void validatePropertyValue(String propertyName, ExtendedConfigDescriptio
...
-		// TODO validate boolean, int, enums types
-		String type = metadata.getType();
+		String value = property.getValue().getText().trim();
+		if (value.length() == 0) {
+			return;
+		}
+		
+		String errorMessage = null;
+		if (isInvalidEnumValue(metadata, value)) {
+			errorMessage = "Invalid enum value: '" + property.getValue().getText().trim() + "' is invalid for type " + metadata.getType();
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java]: replace with 

`errorMessage = "Invalid enum value: '" + value+ "' is invalid for type " + metadata.getType();`

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java [hunk 5]
```diff
@@ -130,13 +130,61 @@ private void validatePropertyValue(String propertyName, ExtendedConfigDescriptio
...
+		} else if (isValueTypeMismatch(metadata, value)) {
+			errorMessage = "Type mismatch: " + metadata.getType() + " expected";
+		}
+
+		if (errorMessage != null) {
+			addDiagnostic(errorMessage, property.getValue(), severity, ValidationType.value.name());
+		}
+	}
+
+	private boolean isInvalidEnumValue(ExtendedConfigDescriptionBuildItem metadata, String value) {
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java]: it's an utility method, please use static

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java [hunk 6]
```diff
@@ -130,13 +130,61 @@ private void validatePropertyValue(String propertyName, ExtendedConfigDescriptio
...
+		for (String e: enums) {
+			if (e.equals(value)) {
+				return false;
+			}
 		}
+
+		return true;
+	}
+
+	private boolean isValueTypeMismatch(ExtendedConfigDescriptionBuildItem metadata, String value) {
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java]: it's an utility method, please use static

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/commons/ExtendedConfigDescriptionBuildItem.java [hunk 7]
```diff
@@ -125,7 +125,19 @@ public boolean isAvailableAtRun() {
...
+	}
+	public boolean isIntegerType() {
+		return "int".equals(getType()) ||
+		"java.lang.Integer".equals(getType()) ||
+		"java.util.OptionalInt".equals(getType()) ||
+		"java.util.Optional<java.lang.Integer>".equals(getType());
+	}
+
+	public boolean isFloatType() {
+		return "float".equals(getType());
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/commons/ExtendedConfigDescriptionBuildItem.java]: It should be:

```java
return "float".equals(getType()) ||
		"java.lang.Float".equals(getType()) ||
		"java.util.Optional<java.lang.Float>".equals(getType());
```

Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/commons/ExtendedConfigDescriptionBuildItem.java]: It misses isLongType, isDoubleType

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java [hunk 8]
```diff
@@ -130,13 +130,61 @@ private void validatePropertyValue(String propertyName, ExtendedConfigDescriptio
...
+		return !isBuildtimePlaceholder(value) &&
+		((metadata.isIntegerType() && !isIntegerString(value)) ||
+		(metadata.isFloatType() && !isFloatString(value)) ||
+		(metadata.isBooleanType() && !isBooleanString(value)));
+	}
+
+	private boolean isBooleanString(String str) {
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java]: it's an utility method, please use static

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java [hunk 9]
```diff
@@ -130,13 +130,61 @@ private void validatePropertyValue(String propertyName, ExtendedConfigDescriptio
...
+		return "true".equals(str) || "false".equals(str);
+	}
+
+	private boolean isIntegerString(String str) {
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java]: it's an utility method, please use static

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java [hunk 10]
```diff
@@ -130,13 +130,61 @@ private void validatePropertyValue(String propertyName, ExtendedConfigDescriptio
...
+		return str.matches("[+-]?\\d+");
+	}
+
+	private boolean isFloatString(String str) {
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java]: it's an utility method, please use static

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java [hunk 11]
```diff
@@ -130,13 +130,61 @@ private void validatePropertyValue(String propertyName, ExtendedConfigDescriptio
...
+		return str.matches("[+-]?(\\d+\\.)?\\d+f?");
+	}
+
+	private boolean isBuildtimePlaceholder(String str) {
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java]: it's an utility method, please use static and please add javadoc

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java [hunk 12]
```diff
@@ -130,13 +130,61 @@ private void validatePropertyValue(String propertyName, ExtendedConfigDescriptio
...
+		return str.matches("^\\$\\{[A-Za-z\\d._-]*\\}$");
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java]: I think we could avoid using regexp (which can be slow) by writting str.startsWith("${");

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java (hunk 11 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java]: what about isLongString and isDoubleString?

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java [hunk 13]
```diff
@@ -130,13 +130,61 @@ private void validatePropertyValue(String propertyName, ExtendedConfigDescriptio
...
+		(metadata.isFloatType() && !isFloatString(value)) ||
+		(metadata.isBooleanType() && !isBooleanString(value)));
+	}
+
+	private boolean isBooleanString(String str) {
+		return "true".equals(str) || "false".equals(str);
+	}
+
+	private boolean isIntegerString(String str) {
+		return str.matches("[+-]?\\d+");
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java]: To improve performance it's better to precompile the regexp. See the following code:

```
	private static final Pattern INTEGER_PATTER = Pattern.compile("[+-]?\\d+");
....	
	private boolean isIntegerString(String str) {
		return INTEGER_PATTER.matcher(str).matches();
	}

```

Please do like this for other matches.

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java (hunk 5 above)
Comment (xorye) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java]: just to confirm, should it be public static or private static?

Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java]: private static as it is used only in this code.  If we see that this method will be used in the future in another code, we could create an utility class with static public.

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java (hunk 13 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java]: BUT I think you should manage this method with Integer.parseInt loike https://stackoverflow.com/questions/237159/whats-the-best-way-to-check-if-a-string-represents-an-integer-in-java

File: quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesDiagnosticsTest.java (hunk 1 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesDiagnosticsTest.java]: yes please

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java (hunk 5 above)
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java]: Once https://github.com/redhat-developer/quarkus-ls/pull/100 will be accepted, remove this method and use `metadata.getEnumItem(value) != null`

Comment (xorye) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusValidator.java]: I think we still need `isInvalidEnumValue(metadata, value)`, because it also checks if the value is valid enum for the metadata or not

**Comment by xorye** [comment]
@angelozerr I've rebased with the branch for this PR: https://github.com/redhat-developer/quarkus-ls/pull/100 and made the changes

**Comment by angelozerr** [comment]
@xorye now #100 was merged in master could you rebase your PR from master please.
